Create proc/func/trigger not visible in the pg_stat_statements view

kritika-0601 opened the pull request:
### Description

[Describe what this change achieves - Guidelines below (please delete the guidelines after writing the PR description)]

Currently Babelfish does not show `create proc/func/trigger` query inside the pg_stat_statements table. After my change now it will be visible in the view.
pg_stat_statements displays planning and execution statistics for all SQL statements executed by a server. Some statements like create proc/func was not being tracked. Mainly because subsequent processutility hooks after bbf_processutilty were not being called. 
Created a hook for that and called it inside relevant ProcessUtility function.

PR for engine code: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/113

### Issues Resolved

[List any issues this PR will resolve]

### Test Scenarios Covered ###
* **Use case based -**


* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'BABEL_3_X_DEV' into pg_stat_statements-kritika

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

kritika-0601 committed:
Merge branch 'BABEL_3_X_DEV' into pg_stat_statements-kritika

Commit:
retrigger tests

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-kritika

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -65,12 +65,18 @@
 #include "multidb.h"
 #include "tsql_analyze.h"
 
+#include "tcop/cmdtag.h"
+#include "tcop/tcopprot.h"
```
Let's move these above with the other includes in alphabetical order. Note, first comes all the engine headers and then the extensions' headers.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -293,10 +310,18 @@ InstallExtendedHooks(void)
+
 	prev_fill_missing_values_in_copyfrom_hook = fill_missing_values_in_copyfrom_hook;
 	fill_missing_values_in_copyfrom_hook = fill_missing_values_in_copyfrom;
+
+
+
```
Nit: unnecessary line breaks.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -293,10 +310,18 @@ InstallExtendedHooks(void)
+
 	prev_check_rowcount_hook = check_rowcount_hook;
 	check_rowcount_hook = bbf_check_rowcount_hook;
+
```
Nit: not needed.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -347,6 +373,169 @@ UninstallExtendedHooks(void)
  * 			Hook Functions
  *****************************************/
 
+void pltsql_CreateFunctionStmt(ParseState *pstate,
```
Nit: missing required line break after `void`. Also, the function needs to be `static`.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -347,6 +373,169 @@ UninstallExtendedHooks(void)
+							   const char *queryString,
+							   bool readOnlyTree,
+							   ProcessUtilityContext context,
+								ParamListInfo params,
```
Nit: Indentation look a bit off.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -347,6 +373,169 @@ UninstallExtendedHooks(void)
+void pltsql_CreateFunctionStmt(ParseState *pstate,
+							   PlannedStmt *pstmt,
+							   const char *queryString,
+							   bool readOnlyTree,
```
Looks redundant as we are always passing `readOnlyTree = false`.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -347,6 +373,169 @@ UninstallExtendedHooks(void)
+							   bool readOnlyTree,
+							   ProcessUtilityContext context,
+								ParamListInfo params,
+							   QueryEnvironment *queryEnv,
+							   DestReceiver *dest,
+							   QueryCompletion *qc)
```
All three look redundant as we are not using them anywhere.

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
retrigger tests

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

rishabhtanwar29 commented:
One other we could do this without introducing a new hook is that tweak the logic in bbf_ProcessUtility itself:
```
// create table type if needed
if (tbltypStmt)
   *create table type*

// call standard utility to create the function.
if (prev_ProcessUtility)
    prev_ProcessUtility(pstmt, queryString, readOnlyTree, context,
						params, queryEnv, dest,
						qc);
else
    standard_ProcessUtility(pstmt, queryString, readOnlyTree, context,
						params, queryEnv, dest,
						qc);

//get address of the function since it has already been created.

// create trigger if needed
if (trigstmt)
   *create the trigger*
```
@2jungkook do you also agree with this?

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

HarshLunagariya commented:
> One other we could do this without introducing a new hook is that tweak the logic in bbf_ProcessUtility itself:
> 
> ```
> // create table type if needed
> if (tbltypStmt)
>    *create table type*
> 
> // call standard utility to create the function.
> if (prev_ProcessUtility)
>     prev_ProcessUtility(pstmt, queryString, readOnlyTree, context,
> 						params, queryEnv, dest,
> 						qc);
> else
>     standard_ProcessUtility(pstmt, queryString, readOnlyTree, context,
> 						params, queryEnv, dest,
> 						qc);
> 
> //get address of the function since it has already been created.
> 
> // create trigger if needed
> if (trigstmt)
>    *create the trigger*
> ```
> 
> @2jungkook do you also agree with this?

There are two problems with above solution:
1. If we do it by calling standard_ProcessUtility() for creating the function only as suggested above then statistics will be only collected for actual create physical function logic. If won't include the statistics for `create table type` or `create trigger` statements. This the main pain point why we can't avoid the hooks here.

2. Even if we go ahead with above approach, then getting the address of function(because we using it in the subsequent logic), we need to rebuild the function signature from the CreateFunctionStmt and search in the catalog.

IMO, To address 1st point for correctness of statistics we should go ahead with hooks approach only.

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'BABEL_3_X_DEV' into pg_stat_statements-kritika

kritika-0601 committed:
Update pl_handler.c

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
retrigger tests

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Merge branch 'BABEL_3_X_DEV' into pg_stat_statements-kritika

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-kritika

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Merge branch 'pg_stat_statements-kritika' of https://github.com/kritika-0601/babelfish_extensions into pg_stat_statements-kritika

Commit:
retrigger tests

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into pg_stat_statements-kritika
